runtime/common/metering.go

```diff
@@ -31,6 +31,8 @@ type MemoryUsage struct {
 	Amount uint64
 }
 
+// MemoryGauge may be used in parallel. e.g: During value encoding.
+// Hence, any implementation of this interface must be thread-safe.
```

Comment:
We should maybe check if is the case in FVM

Reply:
Its not, it uses a map! But it can be made that way. This will have to be a FVM change.

Reply:
If metering is multi-threaded is determinism still guaranteed? 

Is it possible that because things will be metered in random order during encoding that and execution fork will happen because some transactions will fail differently then others? We should make sure that is not the case.

Reply:
hmm, that's a good point. Yes, the order would not be deterministic. All forks will fail eventually, but may not be at the same exact moment. Is that still going to be a problem?

Reply:
for memory measurements no, because the only relevant factor there is if it reaches the limit or not, and that is not dependent on the order. Same for interaction measurements.

Computation measurements are different, as those get computed into fees. 

To illustrate the problem, lets say there are 2 values to encode V1 and V2 at the end of the transaction. both use so much memory that the transaction will fail, but V1 uses 100 computation while v2 uses 200 computation. If V1 is encoded first the transaction will fail and the fees will be `F(100)`, but if V2 is encoded first the transaction will still fail but the fees will be `F(200)` which would cause an execution fork.

Reply:
As discussed, moved the metering to "before" encoding:https://github.com/onflow/cadence/pull/2339/commits/0cbfe2e89a0786b2cbb27c8d976a899857f8a44f

This also made it simpler, and also the memory meter doesn't need to be thread-safe anymore 👍